python/orca/src/bigdl/orca/learn/mpi/mpi_estimator.py

```diff
@@ -308,62 +313,137 @@ def create_plasma_dataloader(config):
     return create_plasma_dataloader
 
 
-def train_epoch(config, model, train_ld, train_batches, optimizer, loss, scheduler,
-                validate_func, valid_ld, metrics, validate_batches, validate_steps):
+def train(config, epochs, model, train_ld, train_batches, optimizer, loss, scheduler,
+          validate_func, valid_ld, metrics, validate_batches, validate_steps):
+    import torch
     import time
     total_loss = 0
     total_samp = 0
     total_iter = 0
     total_time = 0
     previous_iteration_time = None
-    train_iter = iter(train_ld)
-    for j in range(train_batches):
-        # Iterate again from the beginning if running out of batches.
-        if j > 0 and j % len(train_ld) == 0:
-            train_iter = iter(train_ld)
-        current_time = time.time()
-        if previous_iteration_time:
-            iteration_time = current_time - previous_iteration_time
-        else:
-            iteration_time = 0
-        previous_iteration_time = current_time
-        x, y = next(train_iter)
-        o = model(x, y)
-        l = loss(o, y)
-        l_np = l.detach().cpu().numpy()
-        y_np = y.detach().cpu().numpy()
-        optimizer.zero_grad()
-        l.backward()
-        optimizer.step()
-        if scheduler:
-            scheduler.step()
-
-        batch_samples = y_np.shape[0]
-        total_time += iteration_time
-        total_loss += l_np * batch_samples
-        total_iter += 1
-        total_samp += batch_samples
-
-        should_print = ("print_freq" in config and ((j + 1) % config["print_freq"] == 0)) \
-            or (j + 1 == train_batches)
-        if should_print:
-            average_batch_time = 1000.0 * total_time / total_iter
-            total_time = 0
-            average_loss = total_loss / total_samp
-            total_loss = 0
-
-            print(
-                "Finished training it {}/{} of epoch {}, {:.2f} ms/it, ".format(
-                    j + 1, train_batches, config["epoch"], average_batch_time)
-                + "loss {:.6f}, ".format(average_loss)
-            )
-            total_iter = 0
-            total_samp = 0
-
-        should_validate = valid_ld and ((validate_steps > 0 and (j + 1) % validate_steps == 0)
-                                        or (j + 1 == train_batches))
-        if should_validate:
-            validate_func(config, model, valid_ld, metrics, validate_batches)
+    step = 0
+    for i in range(epochs):
```

Comment:
Since you still use the for loop, why do you need to change train_loop? You can put step in config after train_loop?

Reply:
what do you mean? step is to store current step. Use current step to check whether need to print/evaluate when print_frequency/val_step is met.

Reply:
I mean seems you can just keep train_epoch, why do you want users to override the whole train logic?

Reply:
if users want to use huggging face trainer, we can not let them to pass their own train_epoch function because they cannot provide such train_epoch function if using huggingface trainer. We let them to pass whole train logic provide more flexibility to them. If they don't pass custom train function, they will use our default whole train loop.

Reply:
alpaca is trained by batch I remember?